### ISSUE TITLE
Training with 1700 images

### ISSUE THREAD

**Issue by matthew-jurewicz** [lucidrains/stylegan2-pytorch#85]
I've tried .25 and .5 augmentation probability, but just outputs blobs after several hours of training. I'm also using 128 image size and 1 attention layer. Any advice?

**Comment by matthew-jurewicz** [comment]
I also see the gradient penalty jump to hundreds or thousands sometimes.

**Comment by lucidrains** [comment]
@matthew-jurewicz Hey Matthew! Are the images public? I'd be willing to download them and try to debug what is happening

**Comment by matthew-jurewicz** [comment]
Really? That's awesome! I sent you an email.

**Comment by matthew-jurewicz** [comment]
Contrastive loss seems to help keep the numbers down, but output is still blobs.

**Comment by lucidrains** [comment]
Can you post a screenshot of what you are seeing?

**Comment by lucidrains** [comment]
Lol, I see your training data! Have you seen https://digital-thinking.de/watchgan-advancing-generated-watch-images-with-stylegans/

**Comment by lucidrains** [comment]
@matthew-jurewicz which version of the library are you on?

**Comment by matthew-jurewicz** [comment]
![5](https://user-images.githubusercontent.com/20233091/86280998-d32d0980-bbaa-11ea-86a7-d9ca1f5cb5f6.jpg)
This 5000 iterations /w cl-reg

**Comment by matthew-jurewicz** [comment]
Looks like contrastive loss might do the trick with more training.

**Comment by matthew-jurewicz** [comment]
I hadn't seen that! That's super cool!

**Comment by matthew-jurewicz** [comment]
I installed with pip this morning, so presumably the latest.

**Comment by lucidrains** [comment]
![1-ema](https://user-images.githubusercontent.com/108653/86281258-1dee5780-bb92-11ea-9d86-61599bf482e0.jpg) i'm only 1k in and i'm seeing a circle around (the watch?)

**Comment by lucidrains** [comment]
when does it collapse for you?

**Comment by lucidrains** [comment]
@matthew-jurewicz one thing is that the augmentation probability should be set low in the beginning, and elevated as the training progresses. i'll get around to automating that

**Comment by matthew-jurewicz** [comment]
Past iterations don't look any better. What parameters are you using?

**Comment by lucidrains** [comment]
just the default, with an `--aug-prob 0.25`

**Comment by matthew-jurewicz** [comment]
I'm using attention. Maybe that's causing trouble?

**Comment by lucidrains** [comment]
@matthew-jurewicz it could be :( let me investigate the attention today. i recently made a change, and I too have noticed some instability in one of the previous runs

**Comment by matthew-jurewicz** [comment]
Okay cool, you rock!

**Comment by lucidrains** [comment]
@matthew-jurewicz ahh you are right, it collapsed for me around 7k
![6-ema](https://user-images.githubusercontent.com/108653/86284126-1c735e00-bb97-11ea-83c1-7c3c26cd5e0d.jpg)

You can leave this open, I'll figure it out within the week! I have faith in all the papers I've been reading that augmentation should work!

**Comment by matthew-jurewicz** [comment]
When you say collapsed, do you mean modal collapse?

**Comment by lucidrains** [comment]
![9](https://user-images.githubusercontent.com/108653/86296108-e2ad5200-bbac-11ea-91e0-ab4a01561e40.jpg)

I think I may have a solution, should know in another half an hour! yup, so the way this works is that the two AIs battle each other, and near the end, one of them wins and the game is over. the collapse i'm referring to is when one of them wins

**Comment by lucidrains** [comment]
@matthew-jurewicz yup, mode collapse could happen if the discriminator wins and the generator basically focuses on generating one thing

**Comment by lucidrains** [comment]
@matthew-jurewicz ah doh, it still collapsed. let me play around with a few more things

**Comment by lucidrains** [comment]
@matthew-jurewicz ok, seeing some better results (this is at 10k)
![10](https://user-images.githubusercontent.com/108653/86307454-a9d0a580-bbcb-11ea-8451-f1c51ba96dc8.jpg)

i added a new feature, so please upgrade first.

```bash
$ stylegan2_pytorch --data ./path/to/data --aug-prob 0.3 --dataset-aug-prob 0.6 --batch-size 5 --gradient-accumulate-every 6
```

**Comment by matthew-jurewicz** [comment]
Well done! I will give it another shot! Thanks for your help!

**Comment by lucidrains** [comment]
![15](https://user-images.githubusercontent.com/108653/86311484-5e22f980-bbd5-11ea-9c5d-6e8173c26fb4.jpg)

15k!

**Comment by matthew-jurewicz** [comment]
That's crazy! Gone are the big data requirements!

**Comment by lucidrains** [comment]
![20-mr](https://user-images.githubusercontent.com/108653/86316442-a2b49200-bbe1-11ea-912a-51c93a6ca44e.jpg)

20k

**Comment by lucidrains** [comment]
![23-mr](https://user-images.githubusercontent.com/108653/86317685-e5c43480-bbe4-11ea-97a2-4483ddb13f89.jpg)

23k

**Comment by lucidrains** [comment]
![29-ema](https://user-images.githubusercontent.com/108653/86323231-55402100-bbf1-11ea-91a0-23577cb7e384.jpg)
29k

**Comment by matthew-jurewicz** [comment]
Collapsed at 8k for me with following parameters:
Augmentation probability .25
Dataset augmentation probability .5
Batch size 8
Gradient accumulation 4
Attention layers 1 & 2
Image size 256

**Comment by lucidrains** [comment]
![45-mr](https://user-images.githubusercontent.com/108653/86371806-3617b280-bc36-11ea-9ae6-60c08e2064fb.jpg)
For me at 45k

**Comment by matthew-jurewicz** [comment]
Maybe larger image size is more unstable?

**Comment by lucidrains** [comment]
@matthew-jurewicz yeah, it could be, I'll retry with 256x256 later this evening!

**Comment by lucidrains** [comment]
these are really good results for 1700 images already :)

**Comment by lucidrains** [comment]
@matthew-jurewicz have you tried 128x128?

**Comment by matthew-jurewicz** [comment]
Also, does half precision train faster?

**Comment by matthew-jurewicz** [comment]
I haven't tried, I'm paying to run GPU, and I can't really use 128.

**Comment by lucidrains** [comment]
Yup, you can try half precision, but you need to have Nvidia's apex installed

**Comment by lucidrains** [comment]
Ohh why not for 128x128? Also, just letting you know, you can always use Google Colab for free GPUs

**Comment by matthew-jurewicz** [comment]
Okay cool! I need as much detail as possible, so 256 is a good trade-off between detail and training time.

**Comment by matthew-jurewicz** [comment]
Installed Apex, but Nan exception keeps restarting training.

**Comment by matthew-jurewicz** [comment]
Probably something to do with gradient overflow warning.

**Comment by lucidrains** [comment]
@matthew-jurewicz ok, I'll give 256x256 later today once my GPUs free up!

**Comment by matthew-jurewicz** [comment]
Another question, does one iteration represent one epoch?

**Comment by lucidrains** [comment]
@matthew-jurewicz one iteration is just the number on the progress bar, which represents `batch_size * gradient_accumulate_every` samples

**Comment by lucidrains** [comment]
@matthew-jurewicz also, i'm not much of a watch guy, but why do advertised watches all get set to the same time lol?

**Comment by matthew-jurewicz** [comment]
Haha, I didn't notice that. Maybe that's the best looking position according to market research?

**Comment by matthew-jurewicz** [comment]
Or maybe a secret check mark, like "Yes, buy me!"

**Comment by lucidrains** [comment]
@matthew-jurewicz they must have A/B tested the heck out of consumers to settle on that time lol

**Comment by rockdrigoma** [comment]
@lucidrains that's the best position so the brand and the whole clock face is visible

**Comment by andersonfaaria** [comment]
I'm facing a similar problem, the outputs ever since the beginning of trainning are quite strange.
I've tried to run it multiple times, varying parameters and it didn't helped at all.

Those results were achieved using the parameters below, after 9h of trainning.
`!stylegan2_pytorch --new --data "/content/drive/My Drive/stylegan2-colab/stylegan2-1/data/armor" --transparent --aug-prob 0.25 --attn-layers [1,2]`
![image](https://user-images.githubusercontent.com/8560144/89670541-c5f4f000-d8b7-11ea-888d-113d2ed93976.png)

Results achieved by data augmentation tensorflow version, in 2h of trainning:
![image](https://user-images.githubusercontent.com/8560144/89670784-23893c80-d8b8-11ea-932e-dd9bb43ac463.png)

@lucidrains I tried to do a thing similar to you by improving a single repository with all the recent discoveries but I've failed. I was really excited when I found your repo because for projection it requires fp16 model and my dataset is 32x32 with ~100 images so data augmentation and transparency is a must for me. If you could take a look on why the results are so bad in comparison to regular tf implementation and also provide a explanation on how to use fp16 it would be awesome.

**Comment by matthew-jurewicz** [comment]
You probably need 1000 images minimum, try additionally using --dataset-aug-prob .5

**Comment by andersonfaaria** [comment]
> You probably need 1000 images minimum, try additionally using --dataset-aug-prob .5

I'm already using --aug-prob 0.25, this is another parameter that should be used along? There's no mention to that in the readme...
At least from the data augment paper, they tested it with ~100 samples so idk why this isn't working for this version. Also by the images it seems that the data augmentation is leaking

**Comment by matthew-jurewicz** [comment]
That's correct, there's another parameter not specified in the docs which was specifically created for this issue.

**Comment by andersonfaaria** [comment]
you're right, I just noted the comment. I'll try with this new parameter and update my comment here afterwards. Ty

**Comment by matthew-jurewicz** [comment]
Also, nine hours seems like a long time for something small, Google colab has GPU's for free.

**Comment by andersonfaaria** [comment]
I don't know about this version, but regular stylegan2 detects que size of training data and corrects the output version to fit same size. In this version do I need to determine what's the output size or it's by default the same as input?

**Comment by matthew-jurewicz** [comment]
Not sure, the only thing I wrote is this issue lol

**Comment by andersonfaaria** [comment]
Update: I'm at 7k (near 3h trainning) and the model is nowhere near start converging as tf version did. Can you please take a look @lucidrains?

**Comment by matthew-jurewicz** [comment]
I'm trying the same thing with 280, 128 by 128 images

**Comment by matthew-jurewicz** [comment]
I'll say I don't usually notice something until 20000 iterations with batch size 8 and gradient accumulation 4.

**Comment by andersonfaaria** [comment]
I think I'll stick with tf version for now, the results are incomparable atm but thank you for the willingness to help

**Comment by matthew-jurewicz** [comment]
If you're referring to Nvidia repo, it's for non-commercial use only.

**Comment by AlexTS1980** [comment]
I got some pretty exciting results on a very small dataset by increasing the truncation parameter to 0.9 (mode collapses after about 17 epochs)

**Comment by matthew-jurewicz** [comment]
I hadn't noticed that, any idea what it does?

**Comment by AlexTS1980** [comment]
> I hadn't noticed that, any idea what it does?

In the 2018 article it's the equation in Appendix B. Values close to 0s scale the output of W network closer to the conditional expectation. By default it is 0.6, in my case (<1K images data) this quickly (10-12K epochs) led to a mode collapse. Using higher values (e.e 0.9) kept the model evolving varied images up until 18-20K epochs.

**Comment by matthew-jurewicz** [comment]
For anyone curious, I get good results with the following parameters with as little as 200 images.
--aug-prob .25 --dataset-aug-prob .5 --trunc-psi 1 --cl-reg

**Comment by woctezuma** [comment]
For information:

`--aug-prob`

https://github.com/lucidrains/stylegan2-pytorch/blob/0c9b33c3f9e0ee9aec28a0ab8330a8037e802584/stylegan2_pytorch/stylegan2_pytorch.py#L289-L298

`--dataset-aug-prob`

https://github.com/lucidrains/stylegan2-pytorch/blob/0c9b33c3f9e0ee9aec28a0ab8330a8037e802584/stylegan2_pytorch/stylegan2_pytorch.py#L785-L793

https://github.com/lucidrains/stylegan2-pytorch/blob/0c9b33c3f9e0ee9aec28a0ab8330a8037e802584/stylegan2_pytorch/stylegan2_pytorch.py#L720

https://github.com/lucidrains/stylegan2-pytorch/blob/0c9b33c3f9e0ee9aec28a0ab8330a8037e802584/stylegan2_pytorch/stylegan2_pytorch.py#L248-L255

**Comment by JoMe2704** [comment]
> I could not find information in the README about this new command-line argument, so I looked at the code.

As far as I can see, an important difference between --aug-prob and --dataset-aug-prob is that --aug-prob controls the probability of horizontal flipping only (and the other augmentations in AugWrapper, i.e. DiffAugment, ignore --aug-prob), whereas --dataset-aug-prob controls the probability of applying RandomResizedCrop (cropping and subsequent scaling).
